jerry-core/ecma/builtin-objects/ecma-builtin-object.c

```diff
@@ -135,6 +135,141 @@ ecma_builtin_object_object_get_prototype_of (ecma_value_t this_arg, /**< 'this'
 
+#ifndef CONFIG_DISABLE_ES2015_BUILTIN
+/**
+ * [[SetPrototypeOf]]
+ *
+ * See also:
+ *          ES2015 9.1.2
+ */
+static bool
+ecma_set_prototype_of (ecma_value_t o_value, /**< O */
+                       ecma_value_t v_value) /**< V */
+{
+  /* 1. */
+  JERRY_ASSERT (ecma_is_value_object (o_value));
+  JERRY_ASSERT (ecma_is_value_object (v_value) || ecma_is_value_null (v_value));
+
+  ecma_object_t *o_p = ecma_get_object_from_value (o_value);
+
+  /* 3. */
+  ecma_object_t *current_p = ecma_get_object_prototype (o_p);
+  ecma_value_t current_value = ecma_make_object_value (current_p);
+
+  /* 4. */
+  if (ecma_op_same_value (v_value, current_value))
+  {
+    return true;
+  }
+
+  /* 2., 5. */
+  if (!ecma_get_object_extensible (o_p))
+  {
+    return false;
+  }
+
+  if (!ecma_is_value_null (v_value))
+  {
+    ecma_object_t *v_p = ecma_get_object_from_value (v_value);
+
+    /* 6., 7., 8. */
+    ecma_object_t *p_p = v_p;
+    while (true)
+    {
+      /* a. */
+      if (p_p == NULL)
+      {
+        break;
+      }
+
+      /* b. */
+      ecma_value_t p_value = ecma_make_object_value (p_p);
+      if (ecma_op_same_value (p_value, o_value))
+      {
+        return false;
+      }
+
+      /* c.i. TODO: es2015-subset profile does not support having a different
+       * [[GetPrototypeOf]] internal method */
+
+      /* c.ii. */
+      p_p = ecma_get_object_prototype (p_p);
+    }
+
+    /* 9. */
+    ECMA_SET_POINTER (o_p->prototype_or_outer_reference_cp, v_p);
+  }
+  else
+  {
+    /* 9. */
+    ECMA_SET_POINTER (o_p->prototype_or_outer_reference_cp, NULL);
+  }
+
+  /* 10. */
+  return true;
+} /* ecma_set_prototype_of */
+
+/**
+ * The Object object's 'setPrototypeOf' routine
+ *
+ * See also:
+ *          ES2015 19.1.2.18
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_object_object_set_prototype_of (ecma_value_t this_arg, /**< 'this' argument */
+                                             ecma_value_t arg1, /**< routine's first argument */
+                                             ecma_value_t arg2) /**< routine's second argument */
+{
+  JERRY_UNUSED (this_arg);
+  ecma_value_t ret_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
+
+  /* 1., 2. */
+  ECMA_TRY_CATCH (obj_value,
+                  ecma_op_require_object_coercible (arg1),
+                  ret_value);
+
+  /* 3. */
+  if (!ecma_is_value_object (arg2) && !ecma_is_value_null (arg2))
+  {
+    ret_value = ecma_raise_type_error (ECMA_ERR_MSG ("proto is neither Object nor Null."));
+  }
+  else
+  {
+    /* 4. */
+    if (!ecma_is_value_object (obj_value))
+    {
+      ret_value = ecma_copy_value (obj_value);
+    }
+    else
+    {
+      /* 5. */
+      bool status = ecma_set_prototype_of (obj_value, arg2);
+
+      /* 6. TODO: es2015-subset profile does not support having a different
+       * [[SetPrototypeOf]] internal method */
```

Comment:
no. the spec of `Object.setPrototypeOf ( O, proto )` at ES2015 19.1.2.18 (http://www.ecma-international.org/ecma-262/6.0/#sec-object.setprototypeof) prescribes the following steps:

5. Let `status` be `O.[[SetPrototypeOf]](proto)`.
6. `ReturnIfAbrupt(status)`.

but we only have the non-specific `[[SetPrototypeOf]]` implementation right now, which cannot return anything but true/false (i.e., it cannot signal any abortion, cannot throw). thus, step 6 is a no-operation.

as soon as there is a more complex logic in step 5 and we can call something different than `ecma_set_prototype_of`, we will have to flesh out step 6